shared/domains/chelonia/persistent-actions.js

```diff
@@ -0,0 +1,203 @@
+'use strict'
+
+import sbp from '@sbp/sbp'
+
+type SbpInvocation = any[]
+
+type PersistentActionOptions = {
+  errorInvocation?: SbpInvocation,
+  // Maximum number of tries, default: Infinity.
+  maxAttempts: number,
+  // How many seconds to wait between retries.
+  retrySeconds: number,
+  skipCondition?: SbpInvocation,
+  // SBP selector to call on success with the received value.
+  successInvocationSelector?: string,
+  totalFailureInvocation?: SbpInvocation
+}
+
+type PersistentActionStatus = {
+  cancelled: boolean,
+  failedAttemptsSoFar: number,
+  lastError: string,
+  nextRetry: string,
+  pending: boolean
+}
+
+const defaultOptions: PersistentActionOptions = {
+  maxAttempts: Number.POSITIVE_INFINITY,
+  retrySeconds: 30
+}
+const tag = '[chelonia.persistentActions]'
+
+class PersistentAction {
+  invocation: SbpInvocation
+  options: PersistentActionOptions
+  status: PersistentActionStatus
+  timer: TimeoutID | void
+
+  constructor (invocation: SbpInvocation, options: PersistentActionOptions = {}) {
+    this.invocation = invocation
+    this.options = { ...defaultOptions, ...options }
+    this.status = {
+      cancelled: false,
+      failedAttemptsSoFar: 0,
+      lastError: '',
+      nextRetry: '',
+      pending: false
+    }
+  }
+
+  // Do not call if the action is pending or cancelled!
+  async attempt (): Promise<void> {
+    if (await this.trySBP(this.options.skipCondition)) {
+      this.cancel()
+      return
+    }
+    try {
+      this.status.pending = true
+      const result = await sbp(...this.invocation)
+      this.handleSuccess(result)
+    } catch (error) {
+      this.handleError(error)
+    }
+  }
+
+  cancel (): void {
+    this.timer && clearTimeout(this.timer)
+    this.status.cancelled = true
+    this.status.nextRetry = ''
+  }
+
+  async handleError (error: Error): Promise<void> {
+    const { options, status } = this
+    // Update relevant status fields before calling any optional selector.
+    status.failedAttemptsSoFar++
+    status.lastError = error.message
+    const anyAttemptLeft = options.maxAttempts > status.failedAttemptsSoFar
+    status.nextRetry = anyAttemptLeft && !status.cancelled
+      ? new Date(Date.now() + options.retrySeconds * 1e3).toISOString()
+      : ''
+    // Perform any optional SBP invocation.
+    await this.trySBP(options.errorInvocation)
+    !anyAttemptLeft && await this.trySBP(options.totalFailureInvocation)
+    // Schedule a retry if appropriate.
+    if (status.nextRetry) {
+      // Note: there should be no older active timeout to clear.
+      this.timer = setTimeout(() => this.attempt(), this.options.retrySeconds * 1e3)
+    }
+    status.pending = false
+  }
+
+  async handleSuccess (result: any): Promise<void> {
+    const { status } = this
+    status.lastError = ''
+    status.nextRetry = ''
+    status.pending = false
+    this.options.successInvocationSelector &&
+      await this.trySBP([this.options.successInvocationSelector, result])
```

Comment:
Can we refactor this to use events instead of `options.successInvocationSelector`?

I think it makes sense to add at least two events using `okTurtles.events`:

- `CHEL_PERSISTENT_ACTIONS_SUCCESS` - would be emitted above here
- `CHEL_PERSISTENT_ACTIONS_FAILURE` - would be emitted next to line 82 above, next to the `errorInvocation`, along with the `error` object

And if you think of any others that we should emit, feel free to suggest